File: mRemoteV1/CS/Forms/OptionsForm.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
// End of VB project level imports

using mRemoteNG.Forms.OptionsPages;
//using mRemoteNG.App.Runtime;
using mRemoteNG.My;


namespace mRemoteNG.Forms
{
	public partial class OptionsForm
		{
#region Constructors
			public OptionsForm()
			{
				// This call is required by the designer.
				InitializeComponent();
				// Add any initialization after the InitializeComponent() call.
				
				FontOverride(this);
				
				_pages.Add(new Forms.OptionsPages.StartupExitPage(), new PageInfo());
				_pages.Add(new AppearancePage(), new PageInfo());
				_pages.Add(new TabsPanelsPage(), new PageInfo());
				_pages.Add(new ConnectionsPage(), new PageInfo());
				_pages.Add(new SqlServerPage(), new PageInfo());
				_pages.Add(new UpdatesPage(), new PageInfo());
				_pages.Add(new ThemePage(), new PageInfo());
				_pages.Add(new KeyboardPage(), new PageInfo());
				_pages.Add(new AdvancedPage(), new PageInfo());
				
				_startPage = GetPageFromType(typeof(Forms.OptionsPages.StartupExitPage));
				
				_pageIconImageList.ColorDepth = ColorDepth.Depth32Bit;
				PageListView.LargeImageList = _pageIconImageList;
			}
#endregion
			
#region Public Methods
			public DialogResult ShowDialog(IWin32Window ownerWindow, Type pageType)
			{
				_startPage = GetPageFromType(pageType);
				return ShowDialog(ownerWindow);
			}
#endregion
			
#region Private Fields
			private Dictionary<OptionsPage, PageInfo> _pages = new Dictionary<OptionsPage, PageInfo>();
			private ImageList _pageIconImageList = new ImageList();
			
			private OptionsPage _startPage;
			private OptionsPage _selectedPage = null;
#endregion
			
#region Private Methods
#region Event Handlers
			public void OptionsForm_Load(System.Object sender, EventArgs e)
			{
				foreach (KeyValuePair<OptionsPage, PageInfo> keyValuePair in _pages)
				{
					OptionsPage page = KeyValuePair.Key;
					PageInfo pageInfo = KeyValuePair.Value;
					_pageIconImageList.Images.Add(pageInfo.IconKey, page.PageIcon);
					pageInfo.ListViewItem = PageListView.Items.Add(page.PageName, pageInfo.IconKey);
				}
				
				ApplyLanguage();
				LoadSettings();
				
				ShowPage(_startPage);
			}
			
			public void OptionsForm_FormClosing(System.Object sender, FormClosingEventArgs e)
			{
				if (DialogResult == DialogResult.OK)
				{
					SaveSettings();
				}
				else
				{
					RevertSettings();
				}
			}
			
			public void PageListView_ItemSelectionChanged(System.Object sender, ListViewItemSelectionChangedEventArgs e)
			{
				if (!e.IsSelected)
				{
					return ;
				}
				if (_pages.Count < 1)
				{
					return ;
				}
				OptionsPage page = GetPageFromListViewItem(e.Item);
				if (_selectedPage != page)
				{
					ShowPage(page);
				}
				SelectNextControl(PageListView, true, true, true, true);
			}
			
			public void PageListView_MouseUp(System.Object sender, MouseEventArgs e)
			{
				if (PageListView.SelectedIndices.Count == 0)
				{
					PageInfo pageInfo = _pages[_selectedPage];
					pageInfo.ListViewItem.Selected = true;
				}
				SelectNextControl(PageListView, true, true, true, true);
			}
			
			public void OkButton_Click(System.Object sender, EventArgs e)
			{
				DialogResult = DialogResult.OK;
				Close();
			}
			
			public void CancelButtonControl_Click(System.Object sender, EventArgs e)
			{
				DialogResult = DialogResult.Cancel;
				Close();
			}
#endregion
			
			private void ApplyLanguage()
			{
				Text = Language.strMenuOptions;
				OkButton.Text = Language.strButtonOK;
				CancelButtonControl.Text = Language.strButtonCancel;
				
				foreach (OptionsPage page in _pages.Keys)
				{
					try
					{
						page.ApplyLanguage();
					}
					catch (Exception ex)
					{
						MessageCollector.AddExceptionMessage(message: string.Format("OptionsPage.ApplyLanguage() failed for page {0}.", page.PageName), ex: ex, logOnly: true);
					}
				}
			}
			
			private void LoadSettings()
			{
				foreach (OptionsPage page in _pages.Keys)
				{
					try
					{
						page.LoadSettings();
					}
					catch (Exception ex)
					{
						MessageCollector.AddExceptionMessage(message: string.Format("OptionsPage.LoadSettings() failed for page {0}.", page.PageName), ex: ex, logOnly: true);
					}
				}
			}
			
			private void SaveSettings()
			{
				foreach (OptionsPage page in _pages.Keys)
				{
					try
					{
						page.SaveSettings();
					}
					catch (Exception ex)
					{
						MessageCollector.AddExceptionMessage(message: string.Format("OptionsPage.SaveSettings() failed for page {0}.", page.PageName), ex: ex, logOnly: true);
					}
				}
			}
			
			private void RevertSettings()
			{
				foreach (OptionsPage page in _pages.Keys)
				{
					try
					{
						page.RevertSettings();
					}
					catch (Exception ex)
					{
						MessageCollector.AddExceptionMessage(message: string.Format("OptionsPage.RevertSettings() failed for page {0}.", page.PageName), ex: ex, logOnly: true);
					}
				}
			}
			
			private OptionsPage GetPageFromType(Type pageType)
			{
				foreach (OptionsPage page in _pages.Keys)
				{
					if (page.GetType() == pageType)
					{
						return page;
					}
				}
				return null;
			}
			
			private OptionsPage GetPageFromListViewItem(ListViewItem listViewItem)
			{
				foreach (KeyValuePair<OptionsPage, PageInfo> keyValuePair in _pages)
				{
					OptionsPage page = KeyValuePair.Key;
					PageInfo pageInfo = KeyValuePair.Value;
					if (pageInfo.ListViewItem == listViewItem)
					{
						return page;
					}
				}
				return null;
			}
			
			private void ShowPage(OptionsPage newPage)
			{
				if (_selectedPage != null)
				{
					OptionsPage oldPage = _selectedPage;
					oldPage.Visible = false;
					if (_pages.ContainsKey(oldPage))
					{
						PageInfo oldPageInfo = _pages[oldPage];
						oldPageInfo.ListViewItem.Selected = false;
					}
				}
				
				_selectedPage = newPage;
				
				if (newPage != null)
				{
					newPage.Parent = PagePanel;
					newPage.Dock = DockStyle.Fill;
					newPage.Visible = true;
					if (_pages.ContainsKey(newPage))
					{
						PageInfo newPageInfo = _pages[newPage];
						newPageInfo.ListViewItem.Selected = true;
					}
				}
			}
#endregion
			
#region Private Classes
			private class PageInfo
			{
				public string IconKey {get; set;}
				public ListViewItem ListViewItem {get; set;}
				
				public PageInfo()
				{
					IconKey = Guid.NewGuid().ToString();
				}
			}
#endregion
		}
}

File: mRemoteV1/Forms/OptionsForm.vb
Imports mRemote3G.App
Imports mRemote3G.Forms.OptionsPages

Namespace Forms
    Public Class OptionsForm

#Region "Constructors"

        Public Sub New()
            ' This call is required by the designer.
            InitializeComponent()
            ' Add any initialization after the InitializeComponent() call.

            Runtime.FontOverride(Me)

            _pages.Add(New StartupExitPage, New PageInfo)
            _pages.Add(New AppearancePage, New PageInfo)
            _pages.Add(New TabsPanelsPage, New PageInfo)
            _pages.Add(New ConnectionsPage, New PageInfo)
            _pages.Add(New SqlServerPage, New PageInfo)
            _pages.Add(New UpdatesPage, New PageInfo)
            _pages.Add(New ThemePage, New PageInfo)
            _pages.Add(New AdvancedPage, New PageInfo)

            _startPage = GetPageFromType(GetType(StartupExitPage))

            _pageIconImageList.ColorDepth = ColorDepth.Depth32Bit
            PageListView.LargeImageList = _pageIconImageList
        End Sub

#End Region

#Region "Public Methods"

        Public Overloads Function ShowDialog(ownerWindow As IWin32Window, pageType As Type) As DialogResult
            _startPage = GetPageFromType(pageType)
            Return ShowDialog(ownerWindow)
        End Function

#End Region

#Region "Private Fields"

        Private ReadOnly _pages As New Dictionary(Of OptionsPage, PageInfo)
        Private ReadOnly _pageIconImageList As New ImageList

        Private _startPage As OptionsPage
        Private _selectedPage As OptionsPage = Nothing

#End Region

#Region "Private Methods"

#Region "Event Handlers"

        Private Sub OptionsForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
            For Each keyValuePair As KeyValuePair(Of OptionsPage, PageInfo) In _pages
                Dim page As OptionsPage = keyValuePair.Key
                Dim pageInfo As PageInfo = keyValuePair.Value
                _pageIconImageList.Images.Add(pageInfo.IconKey, page.PageIcon)
                pageInfo.ListViewItem = PageListView.Items.Add(page.PageName, pageInfo.IconKey)
            Next

            ApplyLanguage()
            LoadSettings()

            ShowPage(_startPage)
        End Sub

        Private Sub OptionsForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
            If DialogResult = DialogResult.OK Then
                SaveSettings()
            Else
                RevertSettings()
            End If
        End Sub

        Private Sub PageListView_ItemSelectionChanged(sender As Object, e As ListViewItemSelectionChangedEventArgs) _
            Handles PageListView.ItemSelectionChanged
            If Not e.IsSelected Then Return
            If _pages.Count < 1 Then Return
            Dim page As OptionsPage = GetPageFromListViewItem(e.Item)
            If _selectedPage IsNot page Then
                ShowPage(page)
            End If
            SelectNextControl(PageListView, True, True, True, True)
        End Sub

        Private Sub PageListView_MouseUp(sender As Object, e As MouseEventArgs) Handles PageListView.MouseUp
            If PageListView.SelectedIndices.Count = 0 Then
                Dim pageInfo As PageInfo = _pages(_selectedPage)
                pageInfo.ListViewItem.Selected = True
            End If
            SelectNextControl(PageListView, True, True, True, True)
        End Sub

        Private Sub OkButton_Click(sender As Object, e As EventArgs) Handles OkButton.Click
            DialogResult = DialogResult.OK
            Close()
        End Sub

        Private Sub CancelButtonControl_Click(sender As Object, e As EventArgs) Handles CancelButtonControl.Click
            DialogResult = DialogResult.Cancel
            Close()
        End Sub

#End Region

        Private Sub ApplyLanguage()
            Text = Language.Language.strMenuOptions
            OkButton.Text = Language.Language.strButtonOK
            CancelButtonControl.Text = Language.Language.strButtonCancel

            For Each page As OptionsPage In _pages.Keys
                Try
                    page.ApplyLanguage()
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(
                        String.Format("OptionsPage.ApplyLanguage() failed for page {0}.", page.PageName), ex, , True)
                End Try
            Next
        End Sub

        Private Sub LoadSettings()
            For Each page As OptionsPage In _pages.Keys
                Try
                    page.LoadSettings()
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(
                        String.Format("OptionsPage.LoadSettings() failed for page {0}.", page.PageName), ex, , True)
                End Try
            Next
        End Sub

        Private Sub SaveSettings()
            For Each page As OptionsPage In _pages.Keys
                Try
                    page.SaveSettings()
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(
                        String.Format("OptionsPage.SaveSettings() failed for page {0}.", page.PageName), ex, , True)
                End Try
            Next
        End Sub

        Private Sub RevertSettings()
            For Each page As OptionsPage In _pages.Keys
                Try
                    page.RevertSettings()
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(
                        String.Format("OptionsPage.RevertSettings() failed for page {0}.", page.PageName), ex, , True)
                End Try
            Next
        End Sub

        Private Function GetPageFromType(pageType As Type) As OptionsPage
            For Each page As OptionsPage In _pages.Keys
                If page.GetType() Is pageType Then Return page
            Next
            Return Nothing
        End Function

        Private Function GetPageFromListViewItem(listViewItem As ListViewItem) As OptionsPage
            For Each keyValuePair As KeyValuePair(Of OptionsPage, PageInfo) In _pages
                Dim page As OptionsPage = keyValuePair.Key
                Dim pageInfo As PageInfo = keyValuePair.Value
                If pageInfo.ListViewItem Is listViewItem Then Return page
            Next
            Return Nothing
        End Function

        Private Sub ShowPage(newPage As OptionsPage)
            If _selectedPage IsNot Nothing Then
                Dim oldPage As OptionsPage = _selectedPage
                oldPage.Visible = False
                If _pages.ContainsKey(oldPage) Then
                    Dim oldPageInfo As PageInfo = _pages(oldPage)
                    oldPageInfo.ListViewItem.Selected = False
                End If
            End If

            _selectedPage = newPage

            If newPage IsNot Nothing Then
                newPage.Parent = PagePanel
                newPage.Dock = DockStyle.Fill
                newPage.Visible = True
                If _pages.ContainsKey(newPage) Then
                    Dim newPageInfo As PageInfo = _pages(newPage)
                    newPageInfo.ListViewItem.Selected = True
                End If
            End If
        End Sub

#End Region

#Region "Private Classes"

        Private Class PageInfo
            Public Property IconKey As String
            Public Property ListViewItem As ListViewItem

            Public Sub New()
                IconKey = Guid.NewGuid.ToString()
            End Sub
        End Class

#End Region
    End Class
End Namespace